File: App/src/Adaptive.ReactiveTrader.Client/Repositories/TradeRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reactive.Linq;
using Adaptive.ReactiveTrader.Client.Models;
using Adaptive.ReactiveTrader.Client.ServiceClients;

namespace Adaptive.ReactiveTrader.Client.Repositories
{
    class TradeRepository : ITradeRepository
    {
        private readonly IBlotterServiceClient _blotterServiceClient;
        private readonly ITradeFactory _tradeFactory;

        public TradeRepository(IBlotterServiceClient blotterServiceClient, ITradeFactory tradeFactory)
        {
            _blotterServiceClient = blotterServiceClient;
            _tradeFactory = tradeFactory;
        }

        public IObservable<IEnumerable<ITrade>> GetTrades()
        {
            return _blotterServiceClient.GetTrades()
                .Select(trades => trades.Select(_tradeFactory.Create))
                .Catch(Observable.Return(new ITrade[0]))
                .Repeat()
                .Publish()
                .RefCount();
        }
    }
}

File: src/Adaptive.ReactiveTrader.Client.JS/Repositories/TradeRepository.ts
class TradeRepository implements ITradeRepository {
    private _blotterServiceClient : IBlotterServiceClient;
    private _tradeFactory : ITradeFactory;

    constructor(
        blotterServiceClient: IBlotterServiceClient,
        tradeFactory: ITradeFactory) {
        this._tradeFactory = tradeFactory;
        this._blotterServiceClient = blotterServiceClient;
    }

    public getTrades() {
        return Rx.Observable.defer(()=> this._blotterServiceClient.getTradesStream())
            .select(trades=> this.createTrades(trades))
            .catch(()=> Rx.Observable.return([]))
            .repeat()
            .publish()
            .refCount();
    }

    private createTrades(trades: TradeDto[]): ITrade[] {
         var result = [];

        for (var i = 0; i < trades.length; i++) {
            result[i] = this._tradeFactory.create(trades[i]);
        }

        return result;
    }
} 